Whole Genome Bisulfite Sequencing
=================================

Overview
--------
This example demonstrates the use of our smoothing approach for improving the
signal-to-noise ratio for Whole Genome Bisulfite Sequencing (WGBS) methylation
data.

Analysis
--------
First let's load sample data and take a look at what we have:
```{r data_prep, message=FALSE}
source('../smooth.r')
load(file='../data/sample.rda')

# Methylation reads at each site
head(methylation)
summary(methylation)

# Total number of reads at each site
head(coverage)
summary(coverage)

# Methylation locations (offset in Chromosome)
head(cpgsites)
range(cpgsites)
```

Let's now apply our smoothing method and compare the results.
```{r smoothing}
result = fast.smooth(cpgsites, methylation, coverage)

# down-sampled data
head(result$small)

# after interpolation back up to it's full size
head(result$y)
```

Finally, let's create some simple plots of our results.

```{r visualization}
library(ggplot2)

# plot down-sampled data
dat = stack(as.data.frame(result$small))
scale_factor = (max(cpgsites) - min(cpgsites)) / nrow(result$small)
dat$x <- rep(seq_len(nrow(result$small)), ncol(result$small)) * scale_factor
ggplot(dat, aes(x, values)) + geom_line(aes(colour=ind))

# full-size
dat = stack(as.data.frame(result$y))
dat$x <- rep(seq_len(nrow(result$y)), ncol(result$y))
ggplot(dat, aes(x, values)) + geom_line(aes(colour=ind))
```

Done!

